src/python/pants/backend/explorer/graphql/query/targets.py

```diff
@@ -0,0 +1,155 @@
+
+    def __bool__(self) -> bool:
+        return not (self.alias is None and self.limit is None)
+
+    @staticmethod
+    def filter(
+        query: TargetTypesQuery | None, target_types: Iterable[TargetType]
+    ) -> Iterator[TargetType]:
+        if not query:
+            yield from target_types
+            return
+
+        alias_pattern = query.alias and re.compile(query.alias)
+        count = 0
+        for info in target_types:
+            if alias_pattern and not re.match(alias_pattern, info.alias):
+                continue
+            yield info
+            count += 1
+            if query.limit and count >= query.limit:
+                return
+
+
+@strawberry.input
+class TargetsQuery:
+    specs: Optional[List[str]] = None
+    target_type: Optional[str] = None
+    limit: Optional[int] = None
+
+    def __bool__(self) -> bool:
+        # The `specs` field is not used in the `filter` method.
+        return not (self.target_type is None and self.limit is None)
+
+    @staticmethod
+    def filter(query: TargetsQuery | None, targets: Iterable[TargetData]) -> Iterator[TargetData]:
+        if not query:
+            yield from targets
+            return
+
+        count = 0
+        for data in targets:
+            if query.target_type and data.target.alias != query.target_type:
+                continue
+            yield data
+            count += 1
+            if query.limit and count >= query.limit:
+                return
+
+
+@strawberry.type
+class QueryTargetsMixin:
+    """Get targets related info."""
+
+    @strawberry.field
+    def target_types(
+        self, info: Info, query: Optional[TargetTypesQuery] = None
+    ) -> List[TargetType]:
+        """Get all registered target types that may be used in BUILD files."""
+        request_state = GraphQLContext.request_state_from_info(info)
+        return list(
+            TargetTypesQuery.filter(
+                query,
+                (
+                    TargetType.from_help(info)
+                    for info in request_state.all_help_info.name_to_target_type_info.values()
+                ),
+            )
+        )
+
+    @strawberry.field
+    async def targets(self, info: Info, query: Optional[TargetsQuery] = None) -> List[Target]:
+        """Get all targets defined in BUILD files."""
+        req = GraphQLContext.request_state_from_info(info).product_request
+        specs = (
+            specs_parser.parse_specs(query.specs, convert_dir_literal_to_address_literal=False)
+            if query is not None and query.specs
+            else None
+        )
+        targets: AllTargets | Targets
+        if specs:
+            targets = req(Targets, (specs,))
+        else:
+            targets = req(AllTargets)
+
+        # Peek expects to work with unexpanded targets, but we want to present the expanded set of
+        # targets, so we pretend our targets are unexpanded.
+        all_data = req(TargetDatas, (UnexpandedTargets(targets),))
```

Comment:
Not sure about this... matching `peek`'s output probably makes sense here, as it will expose more data overall?